Add explicit language to direct_post error handling (#91)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

    state=eyJhb...6-sVA
```

> The following is a non-normative example of an Authorization Error Response that is sent via an HTTPS request to the Verifier's Response Endpoint:

Reviewer: ```suggestion
The following is a non-normative example of an Authorization Error Response that is sent as an HTTP POST request to the Verifier's Response Endpoint:
```

Reviewer: weird question, but why is authorization response sent via POST is form-urlencoded and not applicaiton/json..?
I know the direct_post definition says: `The Authorization Response parameters are encoded in the body using the application/x-www-form-urlencoded content type.` but..

Author: That is a separate issue. I would have preferred JSON as well. I don't think we can fix that without normative changes.

Reviewer 2: For the record I think the reason is likely that it matches the format received at the redirect_uri (so being consistent makes it slightly easier for existing implementors), and that generally most/all OAuth2 endpoints expect application/x-www-form-urlencoded input.

```

>   POST /post HTTP/1.1
>     Host: client.example.org
>     Content-Type: application/x-www-form-urlencoded

Reviewer 3: ```suggestion
POST /post HTTP/1.1
Host: client.example.org
Content-Type: application/x-www-form-urlencoded
```

Author: Applied your suggestion. I had to fix that that line indentation in other places too.

>     error=...&
>     error_description=...&

Reviewer 2: Sorry, I took another quick look and spotted two minor things  - it's nice if the examples are actual examples as far as possible (sometimes we use '...' but we should try and only do that for things that are too large to include in full)

```suggestion
    error=invalid_request&
    error_description=unsupported%20client_id_scheme&
```

>     state=eyJhb...6-sVA

Reviewer 3: I believe `state` becomes a MUST in this case. Otherwise, how we would distinguish one error response from another.

Author: `state` is REQUIRED as per OAuth2 if provided in the Authz Request. We have that reference in OID4VP already. So, in that case `state` is REQUIRED.

> Other parameters, such as state or code (from [[RFC6749](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#RFC6749)]), or id_token (from [[OpenID.Core](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#OpenID.Core)]), and iss (from [[RFC9207](https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#RFC9207)]) MAY be included in the response as defined in the respective specifications.

Reviewer 2: I (perhaps incorrectly) assumed @adeinega was talking about requiring `state` in the request. I agree the situation in the response is already clear. I think we could be clearer about the use of state in requests in the direct post case at least, I might raise a separate issue for that.

Reviewer 3: @jogu, I was talking about `state` in the response and specifically, about this type of the error response.

@awoie, yes, you are right, `state` is required in the error response if it was presented in the authz request (1), and the authorization server redirects a user to the client (2) per https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1, but for other cases, RFC 6749 doesn't even mention it, take a quick look at https://datatracker.ietf.org/doc/html/rfc6749#section-5.2.

Reviewer 2: >  for other cases, RFC 6749 doesn't even mention it, take a quick look at https://datatracker.ietf.org/doc/html/rfc6749#section-5.2.

This is the token endpoint response section. state is not required there, that's why it's not mentioned. As you say, the authorization error response section (which is the only one relevant to this discussion) does mention state.

```

> If the Response Endpoint has successfully processed the Authorization Response or Authorization Error Response, it MUST respond with HTTPS status code 200.

Reviewer 2: ```suggestion
If the Response Endpoint has successfully processed the Authorization Response or Authorization Error Response, it MUST respond with HTTP status code 200.
```

The following new parameter is defined for use in the response from the endpoint:

`redirect_uri`:

> : OPTIONAL. A string containing a URI. When this parameter is present the Wallet MUST direct the User Agent to this URI. This allows the Verifier to continue the interaction with the End-User on the device where the Wallet resides after the Wallet has sent the Authorization Response to the Response Endpoint. It can be used by the Verifier to prevent session fixation ((#session_fixation)) attacks. The Response Endpoint MAY return the `redirect_uri` parameter for successful Authorization Responses or for Error Responses.

Reviewer: ```suggestion
: OPTIONAL. String containing a URI. When this parameter is present the Wallet MUST redirect the User Agent to this URI. This allows the Verifier to continue the interaction with the End-User on the device where the Wallet resides after the Wallet has sent the Authorization Response to the Response Endpoint. It can be used by the Verifier to prevent session fixation ((#session_fixation)) attacks. The Response Endpoint MAY return the `redirect_uri` parameter in response to successful Authorization Responses or for Error Responses.
```

a term "Response Endpoint" that is used in other places in the spec, too, feels a weird to me but that's a separate issue..

> : OPTIONAL. Value is a string containing a URI. When this parameter is present the Wallet MUST direct the User Agent to this URI. This allows the Verifier to continue the interaction with the End-User on the device where the Wallet resides after the Wallet has sent the Authorization Response to the Response Endpoint. It can be used by the Verifier to prevent session fixation ((#session_fixation)) attacks. The Response Endpoint MAY return the `redirect_uri` parameter for successful Authorization Responses or for Error Responses.

Reviewer 4: ```suggestion
: OPTIONAL. A string containing a URI. When this parameter is present the Wallet MUST direct the User Agent to this URI. This allows the Verifier to continue the interaction with the End-User on the device where the Wallet resides after the Wallet has sent the Authorization Response to the Response Endpoint. It can be used by the Verifier to prevent session fixation ((#session_fixation)) attacks. The Response Endpoint MAY return the `redirect_uri` parameter for successful Authorization Responses or for Error Responses.
```

> : OPTIONAL. When the `redirect_uri` parameter is used the Wallet MUST direct the User Agent to the redirect URI specified by the `redirect_uri` parameter value. The redirect URI allows the Verifier to continue the interaction with the End-User on the device where the Wallet resides after the Wallet has sent the Authorization Response to the Response Endpoint. It especially enables the Verifier to prevent session fixation ((#session_fixation)) attacks. The Response Endpoint MAY return the `redirect_uri` parameter regardless of whether it receives an Authorization Response or Authorization Error Response.

Reviewer 2: I think we're repeating `redirect_uri` a lot considering this paragraph is the definition of the `redirect_uri` parameter.

```suggestion
: OPTIONAL. Value is a string containing a URI. When this parameter is present the Wallet MUST direct the User Agent to this URI. This allows the Verifier to continue the interaction with the End-User on the device where the Wallet resides after the Wallet has sent the Authorization Response to the Response Endpoint. It can be used by the Verifier to prevent session fixation ((#session_fixation)) attacks. The Response Endpoint MAY return the `redirect_uri` parameter for successful Authorization Responses or for Error Responses.
```
